scripts/dts/gen_defines.py

```diff
@@ -53,6 +60,23 @@ def main():
 
     active_compats = set()
 
+    zdid = 1
+    node_to_zdid = {}
+
+    # Assign a ZDID (Zephyr Device ID) to all nodes before we start
+    # we do this as references to nodes like bus parents or via phandles
+    # may happen before we've iterated the referenced node, so we assign
+    # the ZDID for everyone first.
+    for node in edt.nodes:
+        if node.enabled and node.matching_compat:
+            # Skip 'fixed-partitions' devices since they are handled by
+            # write_flash() and would generate extra spurious #defines
+            if node.matching_compat == "fixed-partitions":
+                continue
+
+            node_to_zdid[node] = zdid
+            zdid = zdid + 1
+
```

Comment:
Could move into a separate function (`_init_zdid()` or the like). Would be nice if we could keep `main()` about the same size forever.